ml-agents-envs/mlagents/envs/timers.py

```diff
@@ -0,0 +1,172 @@
+# # Unity ML-Agents Toolkit
+from time import perf_counter
+
+from contextlib import contextmanager
+from typing import Any, Callable, Dict, Generator, TypeVar
+
+"""
+Lightweight, hierarchical timers for profiling sections of code.
+
+Example:
+
+@timed
+def foo(t):
+    time.sleep(t)
+
+def main():
+    for i in range(3):
+        foo(i + 1)
+    with hierarchical_timer("context"):
+        foo(1)
+
+    print(get_timer_tree())
+
+This would produce a timer tree like
+    (root)
+        "foo"
+        "context"
+            "foo"
+
+The total time and counts are tracked for each block of code; in this example "foo" and "context.foo" are considered
+distinct blocks, and are tracked separately.
+"""
```

Comment:
Maybe it is helpful to point out(or give an example) that @timed is totally equivalent to "with hierarchical_timer()" given a certain condition (when function name equals the name passed in). So that for people who are not familiar with Decorator or doesn't want to go to the implementation, it is easier to understand?